Character Creator contrib

InspectorCaracal opened the pull request:
#### Brief overview of PR changes/additions
This adds a light customization of the `charcreate` command and interface, as well as a character creation menu with extensive functionality examples.

#### Motivation for adding to Evennia
Expanding the contribs!

Griatch commented on evennia/contrib/base_systems/character_creator/README.md:
````diff
@@ -0,0 +1,87 @@
+
+Example:
+```py
+from evennia.contrib.base_systems.character_creator.character_creator import ContribChargenAccount
````
I'm not sure this belongs in `contribs/base_systems`. I can see the argument for this be a sort of 'base system' for character generation, but on the other hand a character generation system is very much a RPG-staple, so wouldn't `contribs/rpg` be more suitable?

Griatch commented on evennia/contrib/base_systems/character_creator/README.md:
````diff
@@ -0,0 +1,87 @@
+
+Lastly, in your `settings.py` file, define `CHARGEN_MENU` to your character creation menu module's location.
+
+Example:
+```py
+CHARGEN_MENU = "evennia.contrib.base_systems.character_creator.example_menu"
+```
````
Since this is a setting only introduced for this contrib, I would suggest making this setting optional, and have the code fall back to the contrib-example if nothing is given. This makes the install shorter and easier to test out. Something like

----

By default, the new `charcreate` command provided by this contrib will use the char-creation menu found in `evennia.contrib.base_systems.character_creator.example_menu`. Look at this to get an idea for how it's set up. To tell the contrib code to use your own menu, point to it with the `CHARGEN_MENU` setting. So if your custom chargen menu nodes are found in `mygame/word/chargen_menu.py`, you'd add: 

    CHARGEN_MENU = "world.chargen_menu"

Griatch commented on evennia/contrib/base_systems/character_creator/character_creator.py:
```diff
@@ -0,0 +1,201 @@
+                account.execute_cmd("ic {}".format(char.key))
+
+        EvMenu(session,
+               settings.CHARGEN_MENU,
```
I'd suggest loading this at the top, with an `AttributeError` leading to falling back to the example menu.

Griatch commented on evennia/contrib/base_systems/character_creator/example_menu.py:
```diff
@@ -0,0 +1,444 @@
+"""
+An example EvMenu for the character creator contrib.
```
I like the instructional nature of this! :+1:

Griatch commented on evennia/contrib/base_systems/character_creator/example_menu.py:
This works fine, but it's worth to note that `evennia.utils.evmenu.list_node` decorator is specifically there to auto-generate a multiple-choice menu node. So it could be worth to point to that since it would lead to a lot less code.

Griatch commented on evennia/contrib/base_systems/character_creator/example_menu.py:
Minor detail - but FYI, if you have only one option, you don't need to use a list here, but can just return the dict as-is. Keeping the list can be said to be more style-consistent though.

Griatch commented on evennia/contrib/base_systems/character_creator/README.md:
````diff
@@ -0,0 +1,87 @@
+In your game folder `commands/default_cmdsets.py`, import and add `ContribCmdCharCreate` to your `AccountCmdSet`.
+
+Example:
+```py
````
Unlike in discord, the full `python` language name should usually be used here.

Griatch commented on evennia/contrib/base_systems/character_creator/example_menu.py:
Why echo the same node? Would it not make more sense to say "This was already taken", and go back to the `menunode_choose_name` node directly - that one likely contains info about the name you can pick (it could fail validation for different reasons after all, such as containing invalid characters etc - things that could be informed of in the first node). 

An alternative to doing two nodes like this would thus be to use a storage-helper (useful for all nodes in the menu) and a validation-helper (for the name node only) and have a single menu node that repeats. This is more a suggestion (if you are not familiar with the pattern) rather than a request for change though. 

```python
def _store_data(caller, raw_string, **kwargs): 
    # generic storage pass-through (useful for all nodes)
    data = kwargs.pop("data", None) 
    next_node = kwargs.pop("next_node", None)  
    if data: 
        character.ndb.chargen_data[data['key']] = data['value']
    return next_node, kwargs 

def _validate_name(caller, raw_string, **kwargs): 
    # validate input 
    if valid: 
        return None, {"name": name}
    else: 
        caller.msg("This name is invalid because ...") 
        return None, {} 

def node_name(caller, raw_string, **kwargs): 

     new_name = kwargs.get("name")

     if new_name: 
         text = "Are you sure?" 
         options = (
                {
                    "key": ("Yes", "y"), 
                    "goto": (_store_data, {"data": {"key": "name", "value": new_name}})
                 },{
                     "key" ("No", "n"),
                     "goto": "node_name"
                }
         )
     else: 
         text = "What name do you want?" 
         options = {"key": "_default",  
                          "goto": _validate_name}
    return text, options

```

Griatch commented on evennia/contrib/base_systems/character_creator/example_menu.py:
One thing (not in the original chargen command either) is the normalization of the name. This is to avoid people making names that are visually similar to other names (by using unicode trickery). The helper is `Account.normalize_username`. May be relevant to add here (and probably to the original chargen command too).

Griatch reviewed (changes requested):
I think this will be an appreciated contrib. I have some suggestions and comments, but overall this looks very useful and informative. :+1:

InspectorCaracal commented:
I've looked over the comments but haven't had a chance to start going through them in detail yet. Just wanted to drop a note that this hasn't fallen off my radar. 👍

Griatch commented:
Ping :)

InspectorCaracal commented:
> Ping :)

Thanks for the ping! I've finally recovered from all the nonsense involved in having family visiting for two weeks, so I'll get through this on Monday. :+1:

InspectorCaracal commented:
Okay, I think that covers all the comments! To summarize:

- moved to `contrib.rpg` (I wasn't sure where to put it, so I was hoping if it was the wrong place you'd suggest something better xD)
- used the example menu as a fallback setting
- rewrote the name selection nodes to be structured more cleanly, I don't know *what* I was thinking with how I did it the first time
- added a comment block mentioning list nodes to the example menu

I also didn't know you could assign a single option dict without putting it in a list! For my own code I like the consistency, but I changed those to just the dictionary of the option for the example, since the point of it is to demonstrate what's possible.

InspectorCaracal commented:
Okay, done with tweaks for real this time.

InspectorCaracal commented:
I was just reminded! This needs to be updated for the new character settings, I will get that patched in today.
